Revit_Core_Engine/Query/OpeningSurface.cs

```diff
@@ -54,12 +63,102 @@ public static ISurface OpeningSurface(this FamilyInstance familyInstance, HostOb
-                hosts = hosts.Where(x => x.FindInserts(true, true, true, true).Any(y => y.IntegerValue == familyInstance.Id.IntegerValue)).ToList();
+                hosts = hosts.Where(x => x.FindInserts(true, true, true, true).Any(y => (y.IntegerValue == familyInstance.Id.IntegerValue)|| (y.IntegerValue == parentId))).ToList();
             }
 
             List<ISurface> surfaces = new List<ISurface>();
 
-            if (hosts.Count == 0)
+            if (parentElem != null) // Indicates element is a nested instance and needs different approach to extract location
+            {
+                Document doc = familyInstance.Document;
+                Transaction t = new Transaction(doc);
+
+                FailureHandlingOptions failureHandlingOptions = t.GetFailureHandlingOptions().SetClearAfterRollback(true);
+                t.Start("Temp Delete Inserts");
+                List<ElementId> inserts = hosts.SelectMany(x => x.FindInserts(true, true, true, true)).Distinct().Where(x => x.IntegerValue != -1).ToList();
+                ParameterSet instanceParams = familyInstance.Parameters;
+                List<RevitParameter> bhomParams = new List<RevitParameter>();
+                foreach (Parameter parameter in instanceParams)
+                {
+                    RevitParameter bhomParam = Convert.ParameterFromRevit(parameter);
+                    bhomParams.Add(bhomParam);
+                }
+                    
+                LocationPoint locPt = familyInstance.Location as LocationPoint;
+                Element dummyInstance = doc.Create.NewFamilyInstance(locPt.Point, familyInstance.Symbol, hosts.First() as Element, Autodesk.Revit.DB.Structure.StructuralType.NonStructural);
+
+                foreach (RevitParameter param in bhomParams.Where(x => !(x.Value is double)).Union(bhomParams.Where(x => x.Value is double)))
+                {
+                    Parameter dummyParam = dummyInstance.GetParameters(param.Name).First<Parameter>();
+                    List<string> paramsExcluded = new List<string> { "Sill Height", "Head Height" };
+                    if (!paramsExcluded.Contains(param.Name))
+                        Modify.SetParameters(dummyInstance, param.Name, param.Value);
+                }
+                doc.Delete(inserts);
+                doc.Regenerate();
+
+                List<List<Solid>> solidsWithOpening = new List<List<Solid>>();
+                foreach (HostObject h in hosts)
+                {
+                    solidsWithOpening.Add(h.Solids(new Options()).Select(x => SolidUtils.Clone(x)).ToList());
+                }
+
+                // Rollback and restart of the transaction is needed because otherwise the object, to which familyInstance is pointing can become invalidated.
+                t.RollBack(failureHandlingOptions);
+                t.Start();
+
+                List<CurveLoop> loops = new List<CurveLoop>();
```

Comment:
This piece of code looks very similar to the existing implementation - couldn't it be wrapped into a private method and called twice to avoid boilerplate?

Reply:
done